src/client.rs

```diff
@@ -1568,6 +1589,29 @@ impl<C: Config> Client<C> {
         self.bits.set_created_this_tick(false);
 
         send_packet(&mut self.send, S2cPlayMessage::Flush);
+
+        // Update the player's inventory
+        if self.inventory.is_dirty() {
+            send_packet(
+                &mut self.send,
+                SetContainerContent {
+                    window_id: 0,
+                    state_id: VarInt(self.inventory.state_id),
+                    slots: self
+                        .inventory
+                        .slots()
+                        .into_iter()
+                        // FIXME: cloning is necessary here to build the packet.
+                        // However, it should be possible to avoid the clone if this packet
+                        // could consume refs
+                        .map(|s| s.cloned())
+                        .collect(),
+                    carried_item: self.cursor_held_item.clone(),
+                },
+            );
+            self.inventory.state_id = self.inventory.state_id.wrapping_add(1);
+            self.inventory.mark_dirty(false);
+        }
```

Comment:
This should happen above where the packets are flushed. Otherwise the packets will arrive a tick late.

Reply:
Just to clarify, is this the line that flushes the packets?

```rust
send_packet(&mut self.send, S2cPlayMessage::Flush);
```

Reply:
Yeah